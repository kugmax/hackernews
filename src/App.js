import React, { Component } from 'react';
import logo from './logo.svg';
import './App.css';

const DEFAULT_QUERY = 'redux';
const DEFAULT_HPP = '100'

const PATH_BASE = 'https://hn.algolia.com/api/v1';
const PATH_SEARCH = '/search';
const PARAM_SEARCH = 'query=';
const PARAM_PAGE = 'page='
const PARAM_HPP = 'hitsPerPage='

const BookNode = ({index, book, onDismissHandler}) =>
  <div key={book.objectID} className="table-row">
    <span style={{width: '2%'}}>
      {index}
    </span>
    <span style={{width: '5%'}}>
      {book.objectID}
    </span>
    <span style={{width: '40%'}}>
      <a target="_blank" href={book.url}>{book.title}</a>
    </span>
    <span style={{width: '20%'}}>
      {book.author}
    </span>
    <span style={{width: '10%'}}>
      {book.num_comments}
    </span>
    <span style={{width: '10%'}}>
      {book.points}
    </span>
    <span style={{width: '10%'}}>
      <Button
        onClick={() => onDismissHandler(book.objectID)}
        className="button-inline">
        Dismiss
      </Button>
    </span>
  </div>

const BooksList = ({list, onDismissHandler}) =>
  list
  .map( (item, index) =>
    <div calssName="table">
      <BookNode index= {index + 1} book={item} onDismissHandler={onDismissHandler}/>
    </div>
  );

const Search = ({value, onChange, onSubmit, children}) =>
  <form onSubmit={onSubmit} className="table-header">
    <input
      type="text"
      value={value}
      onChange={onChange}
      />

    <button type="submit">
      {children}
    </button>
  </form>

const Button = ({onClick, className = '', children}) =>
  <button
    onClick={onClick}
    className={className}
    type="button">
    {children}
  </button>

class App extends Component {
  constructor(props) {
    super(props);
    this.state = {
      result: null,
      searchTerm: DEFAULT_QUERY,
    };
  }

  setSearchTopStories = (result) => {
    console.log(result)

    const {hits, page} = result;
    const oldHits = page !== 0 ? this.state.result.hits : [];
    const updatedHits = [...oldHits, ...hits];

    this.setState({
      result: {hits: updatedHits, page}
    })

  }

  componentDidMount = () => {
    const {searchTerm} = this.state;
    this.fetchSearchTopStories(searchTerm);
  }


  onDismiss = (id) => {
    const isNotID = item => item.objectID != id;
    const updatedHits = this.state.result.hits.filter(isNotID);
    this.setState({
      result: { ...this.state.result, hits: updatedHits}
    })
  }

  onSearchChange = (event) => {
    this.setState({searchTerm: event.target.value})
  }

  onSearchSubmit = (event) => {
    const { searchTerm } = this.state;
    this.fetchSearchTopStories(searchTerm);
    event.preventDefault();
  }

  fetchSearchTopStories = (searchTerm, page=0) => {
    fetch(`${PATH_BASE}${PATH_SEARCH}?${PARAM_SEARCH}${searchTerm}
      &${PARAM_PAGE}${page}&${PARAM_HPP}${DEFAULT_HPP}`)
    .then(response => response.json())
    .then(result => this.setSearchTopStories(result))
    .catch(error => error);
  }

  render() {
    const { searchTerm, result } = this.state;
    const page = (result && result.page) || 0;

    if (!result) {return null;}

    return (
      <div className="page">
        <div className="interactions">
          <Search
            value={searchTerm}
            onChange={this.onSearchChange}
            onSubmit={this.onSearchSubmit}
            >
            Search
          </Search>
        </div>

        {
          result ?
          <BooksList
            list={result.hits}
            onDismissHandler={this.onDismiss}/>
          : null
        }
        <div className="interactions">
          <Button onClick={() => this.fetchSearchTopStories(searchTerm, page+1)}>
            More
          </Button>
        </div>
      </div>
    );
  }
}

export default App;
