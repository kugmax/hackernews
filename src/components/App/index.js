import React, { Component } from 'react';
import {
  DEFAULT_QUERY, DEFAULT_HPP,
  PATH_BASE, PATH_SEARCH, PARAM_SEARCH, PARAM_PAGE, PARAM_HPP
} from '../../constants';
import Button from '../Button';
import Search from '../Search';
import BooksList from '../Table';
import Loading from '../Loading';

import axios from 'axios';
import './index.css';

import { library } from '@fortawesome/fontawesome-svg-core'
import { faSpinner , faArrowUp, faArrowDown} from '@fortawesome/free-solid-svg-icons'

library.add(faSpinner, faArrowUp, faArrowDown );

const withLoading = (Component) => ({ isLoading, ...rest }) =>
  isLoading
  ? <Loading/>
  : <Component {...rest}/>

const ButtonWithLoading = withLoading(Button);

export const updateSearchTopStoriesState = (hits, page) => (prevState) => {
  const { searchKey, results } = prevState;

  const oldHits = results && results[searchKey]
    ? results[searchKey].hits
    : [];
  const updatedHits = [...oldHits, ...hits];

  return {
    results: {
      ...results,
      [searchKey]: {hits: updatedHits, page}
    },
    isLoading: false
  };
};

class App extends Component {
  constructor(props) {
    super(props);
    this.state = {
      results: null,
      searchKey: '',
      searchTerm: DEFAULT_QUERY,
      error: null,
      isLoading: false,
    };
  }

  setSearchTopStories = (json) => {
    const { hits, page } = json;
    this.setState(updateSearchTopStoriesState(hits, page));
  };

  componentDidMount = () => {
    const { searchTerm } = this.state;
    this.setState({ searchKey: searchTerm })
    this.fetchSearchTopStories(searchTerm);
  }

  onDismiss = (id) => {
    const { searchKey, results } = this.state;
    const { hits, page } = results[searchKey];

    const isNotID = item => item.objectID !== id;
    const updatedHits = hits.filter(isNotID);

    this.setState({
      results: {
        ...results,
        [searchKey]: {hits: updatedHits, page}
      }
    })
  }

  onSearchChange = (event) => {
    this.setState({searchTerm: event.target.value})
  }

  needsToSearchTopStories = (searchTerm) => !this.state.results[searchTerm]

  onSearchSubmit = (event) => {
    const { searchTerm } = this.state;
    this.setState({ searchKey: searchTerm })

    if (this.needsToSearchTopStories(searchTerm)) {
      this.fetchSearchTopStories(searchTerm);
    }

    event.preventDefault();
  }

  fetchSearchTopStories = (searchTerm, page=0) => {
    this.setState({ isLoading: true });

    axios(`${PATH_BASE}${PATH_SEARCH}?${PARAM_SEARCH}${searchTerm}
      &${PARAM_PAGE}${page}&${PARAM_HPP}${DEFAULT_HPP}`)
    .then(respons => this.setSearchTopStories(respons.data))
    .catch(error => this.setState({ error }));
  }

  render() {
    const { searchTerm, results, searchKey, error, isLoading} = this.state;
    const page = (results && results[searchKey] && results[searchKey].page) || 0;
    const list = (results && results[searchKey] && results[searchKey].hits) || [];

    return (
      <div className="page">
        <div className="interactions">
          <Search
            value={searchTerm}
            onChange={this.onSearchChange}
            onSubmit={this.onSearchSubmit}
            >
            Search
          </Search>
        </div>
        { error
          ? <div className="interactions">
            <p>Something went wrong.</p>
          </div>
          :
          <BooksList
              list={list}
              onDismissHandler={this.onDismiss}/>
        }
        <div className="interactions">
          <ButtonWithLoading
            isLoading={isLoading}
            onClick={() => this.fetchSearchTopStories(searchKey, page+1)}>
            More
          </ButtonWithLoading>
        </div>
      </div>
    );
  }
}

export default App;

export {
  Button, Search, BooksList
};
