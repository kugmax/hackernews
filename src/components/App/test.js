import React from 'react';
import ReactDOM from 'react-dom';
import App, { Search, Button, BooksList } from './index';
import renderer from 'react-test-renderer'

import Enzyme, { shallow, render, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

Enzyme.configure({ adapter: new Adapter() });

describe('App', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<App />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  test('has a valid snapshot', () => {
      const component = renderer.create(
        <App/>
      )
      const tree = component.toJSON();
      expect(tree).toMatchSnapshot();
  });

});

describe('Search', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<Search>Search</Search>, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  test('has a valid snapshot', () => {
    const component = renderer.create(
      <Search>Search</Search>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('Button', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<Button>Give me more more!</Button>, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  test('has a valid snapshot', () => {
    const component = renderer.create(
      <Button>Give me more</Button>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it ('should use className', () => {
    const element = shallow(
      <Button className='giveMeMore'>Give me more</Button>
    )
    expect(element.find('.giveMeMore').length).toBe(1)
  });
});

describe('BookList', () => {
  const props = {
    list: [
      {title: '1', author: 'Darth', num_comments: 1, points: 2, objectID: 'y'},
      {title: '2', author: 'Mauth', num_comments: 1, points: 2, objectID: 'z'},
    ],
  };

  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<BooksList { ...props }/>, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  test('has a valid snapshot', () => {
    const component = renderer.create(
      <BooksList { ...props }/>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('shows tow items in list', () => {
    const element = mount(
      <BooksList {...props}/>
    );
    expect(element.find('.table').length).toBe(1);
    expect(element.find('.table-row').length).toBe(2);
  });

});
