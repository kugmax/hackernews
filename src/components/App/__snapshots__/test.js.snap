// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App has a valid snapshot 1`] = `
<div
  className="page"
>
  <div
    className="interactions"
  >
    <form
      className="table-header"
      onSubmit={[Function]}
    >
      <input
        onChange={[Function]}
        type="text"
        value="redux"
      />
      <button
        type="submit"
      >
        Search
      </button>
    </form>
  </div>
  <div
    className="table"
  />
  <div
    className="interactions"
  >
    <button
      className=""
      onClick={[Function]}
      type="button"
    >
      More
    </button>
  </div>
</div>
`;

exports[`BookList has a valid snapshot 1`] = `
<div
  className="table"
>
  <div
    className="table-row"
  >
    <span
      style={
        Object {
          "width": "2%",
        }
      }
    >
      1
    </span>
    <span
      style={
        Object {
          "width": "5%",
        }
      }
    >
      y
    </span>
    <span
      style={
        Object {
          "width": "40%",
        }
      }
    >
      <a
        target="_blank"
      >
        1
      </a>
    </span>
    <span
      style={
        Object {
          "width": "20%",
        }
      }
    >
      Darth
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      1
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      2
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      <button
        className="button-inline"
        onClick={[Function]}
        type="button"
      >
        Dismiss
      </button>
    </span>
  </div>
  <div
    className="table-row"
  >
    <span
      style={
        Object {
          "width": "2%",
        }
      }
    >
      2
    </span>
    <span
      style={
        Object {
          "width": "5%",
        }
      }
    >
      z
    </span>
    <span
      style={
        Object {
          "width": "40%",
        }
      }
    >
      <a
        target="_blank"
      >
        2
      </a>
    </span>
    <span
      style={
        Object {
          "width": "20%",
        }
      }
    >
      Mauth
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      1
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      2
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      <button
        className="button-inline"
        onClick={[Function]}
        type="button"
      >
        Dismiss
      </button>
    </span>
  </div>
</div>
`;

exports[`Button has a valid snapshot 1`] = `
<button
  className=""
  type="button"
>
  Give me more
</button>
`;

exports[`Search has a valid snapshot 1`] = `
<form
  className="table-header"
>
  <input
    type="text"
  />
  <button
    type="submit"
  >
    Search
  </button>
</form>
`;
